quickwit/quickwit-serve/src/pprof.rs

```diff
@@ -0,0 +1,106 @@
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use warp::Filter;
+
+/// pprof/start to start cpu profiling
+/// pprof/stop to stop cpu profiling and return a flamegraph
+#[cfg(not(feature = "pprof"))]
+pub fn pprof_routes() -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
+    let start_profiler = {
+        warp::path!("pprof" / "start").map(move || {
+            warp::reply::with_status(
+                "not compiled with pprof feature",
+                warp::http::StatusCode::BAD_REQUEST,
+            )
+        })
+    };
+
+    let stop_profiler = {
+        warp::path!("pprof" / "stop").map(move || {
+            warp::reply::with_status(
+                "not compiled with pprof feature",
+                warp::http::StatusCode::BAD_REQUEST,
+            )
+        })
+    };
+
+    start_profiler.or(stop_profiler)
+}
+
+/// pprof/start to start cpu profiling
+/// pprof/stop to stop cpu profiling and return a flamegraph
+#[cfg(feature = "pprof")]
+pub fn pprof_routes() -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
+    use std::sync::{Arc, Mutex};
+
+    use pprof::ProfilerGuard;
+    use warp::reply::Reply;
+    let profiler_guard: Arc<Mutex<Option<ProfilerGuard<'_>>>> = Arc::new(Mutex::new(None));
+
+    let start_profiler = {
+        let profiler_guard = Arc::clone(&profiler_guard);
+        warp::path!("pprof" / "start").map(move || {
+            let mut guard = profiler_guard.lock().unwrap();
+            if guard.is_none() {
+                *guard = Some(pprof::ProfilerGuard::new(100).unwrap());
+                warp::reply::with_status("CPU profiling started", warp::http::StatusCode::OK)
+            } else {
+                warp::reply::with_status(
+                    "CPU profiling is already running",
+                    warp::http::StatusCode::BAD_REQUEST,
+                )
+            }
+        })
+    };
+
+    let stop_profiler = {
+        let profiler_guard = Arc::clone(&profiler_guard);
+        warp::path!("pprof" / "stop").map(move || {
+            let profiler_guard = Arc::clone(&profiler_guard);
+            get_flamegraph(profiler_guard)
+        })
+    };
+
+    fn get_flamegraph(profiler_guard: Arc<Mutex<Option<ProfilerGuard>>>) -> impl warp::Reply {
```

Comment:
wouldn't the following have worked too?

```suggestion
    fn get_flamegraph(profiler_guard: &Mutex<Option<ProfilerGuard>>) -> impl warp::Reply {
```

Reply:
in the old version yes, in the new version we need a static lifetime